azure-iot-device/tests/iothub/pipeline/test_iothub_pipeline.py

```diff
@@ -7,505 +7,1407 @@
+    def test_op_fail(self, mocker, pipeline, feature):
+        pipeline.disable_feature(feature)
+        op = pipeline._pipeline.run_op.call_args[0][0]
+        op.error = Exception()
+
+        with pytest.raises(SystemExit):
+            op.callback(op)
+
+    @pytest.mark.it(
+        "Does not trigger callback upon unsuccessful completion of the DisableFeatureOperation"
+    )
+    @pytest.mark.parametrize("feature", all_features)
+    def test_op_fail_no_callback(self, mocker, pipeline, feature):
+        cb = mocker.MagicMock()
+        pipeline.disable_feature(feature, callback=cb)
+        op = pipeline._pipeline.run_op.call_args[0][0]
+        op.error = Exception()
 
+        with pytest.raises(SystemExit):
+            op.callback(op)
 
-class TestDisableInputMessage:
-    def test_unsubscribe_of_input_calls_unsubscribe_on_provider(self, module_pipeline_adapter):
-        mock_mqtt_client_operator = module_pipeline_adapter._pipeline.client_operator
+        assert cb.call_count == 0
 
-        module_pipeline_adapter.connect()
-        mock_mqtt_client_operator.on_mqtt_connected()
-        module_pipeline_adapter.disable_feature(constant.INPUT_MSG)
 
-        mock_mqtt_client_operator.unsubscribe.assert_called_once_with(
-            topic=subscribe_input_message_topic, callback=ANY
-        )
+@pytest.mark.describe("IoTHubPipeline - EVENT: Connected")
+class TestIoTHubPipelineEVENTConnect(object):
+    @pytest.mark.it("Triggers the 'on_connected' handler")
+    def test_with_handler(self, mocker, pipeline):
+        # Set the handler
+        mock_handler = mocker.MagicMock()
+        pipeline.on_connected = mock_handler
+        assert mock_handler.call_count == 0
 
-    def test_sets_input_message_status_to_disabled(self, module_pipeline_adapter):
-        mock_mqtt_client_operator = module_pipeline_adapter._pipeline.client_operator
+        # Trigger the connect
+        pipeline._pipeline.on_connected()
 
-        module_pipeline_adapter.connect()
-        mock_mqtt_client_operator.on_mqtt_connected()
-        module_pipeline_adapter.disable_feature(constant.INPUT_MSG)
+        assert mock_handler.call_count == 1
+        assert mock_handler.call_args == mocker.call()
 
-        assert not module_pipeline_adapter.feature_enabled[constant.INPUT_MSG]
+    @pytest.mark.it("Does nothing if the 'on_connected' handler is not set")
+    def test_without_handler(self, pipeline):
+        pipeline._pipeline.on_connected()
 
+        # No assertions required - not throwing an exception means the test passed
 
-class TestEnableC2D:
-    def test_subscribe_calls_subscribe_on_provider(self, device_pipeline_adapter):
-        mock_mqtt_client_operator = device_pipeline_adapter._pipeline.client_operator
 
-        device_pipeline_adapter.connect()
-        mock_mqtt_client_operator.on_mqtt_connected()
-        device_pipeline_adapter.enable_feature(constant.C2D_MSG)
+@pytest.mark.describe("IoTHubPipeline - EVENT: Disconnected")
```

Comment:
>EVENT [](start = 40, length = 5)

Why is this FULL CAPS suddenly ?

Reply:
I so painstakingly wrote a whole series of wiki articles about it and you didn't even read them :cry: :cry: :cry: :cry: 

---
In reply to: [299254539](https://github.com/Azure/azure-iot-sdk-python-preview/pull/120#discussion_r299254539) [](ancestors = 299254539)